core/modules/wbase/SendChannelShared.cc

```diff
@@ -0,0 +1,299 @@
+        // throw Bug doesn't count as it should never happen and will crash the program anyway.
+        if (thisTransmit == nullptr) {
+            throw Bug("_transmitLoop() _transmitQueue had nullptr!");
+        } else if (thisTransmit->result == nullptr) {
+            throw Bug("_transmitLoop() had nullptr result!");
+        }
+        idStr = makeIdStr(thisTransmit->result->queryid(), thisTransmit->result->jobid());
+
+        // Append the header for the next message to  thisTransmit->dataMsg.
+        auto sz = _transmitQueue.size();
+        // Is this really the last message for this SharedSendChannel.
+        bool reallyLast = (_lastRecvd && sz == 0);
+
+        proto::ProtoHeader* nextPHdr;
+        if (sz == 0) {
+            if (!_lastRecvd) {
+                throw Bug("SendChannelShared::_transmit() size=" + to_string(sz) + " but not last.");
+            }
+            // Create a header for an empty result using the protobuf arena from thisTransmit.
+            // This is the signal to the czar that this SharedSendChannel is finished.
+            nextPHdr = thisTransmit->createHeader();
+        } else {
+            auto nextT = _transmitQueue.front();
+            nextPHdr = nextT->header;
+        }
+        nextPHdr->set_endnodata(reallyLast);
+        string nextHeaderString;
+        nextPHdr->SerializeToString(&nextHeaderString);
+        thisTransmit->dataMsg += proto::ProtoHeaderWrap::wrap(nextHeaderString);
+        if (reallyLast) loop = false;
+        // Done with the queue for a while and this function may need to wait below.
+        qLock.unlock();
+        _qBackCv.notify_one();
+
+        // Limit the number of concurrent transmits.
+        bool scanInteractive = thisTransmit->scanInteractive;
+        // If there was an error, give this high priority.
+        if (thisTransmit->erred) {
+            scanInteractive = true;
+        }
+        bool largeResult = thisTransmit->largeResult;
+        int czarId = thisTransmit->czarId;
+        // Do NOT create transmitLock this with a mutex locked. High risk of deadlock.
+        wcontrol::TransmitLock transmitLock(*_transmitMgr, scanInteractive, largeResult, czarId);
+
+        // The first message needs to put its header data in metadata as there's
+        // no previous message it could attach its header to.
+        if (_firstTransmit.exchange(false)) {
+            // Put the header for the first message in metadata
+            // _metaDataBuf must remain valid until Finished() is called.
+            proto::ProtoHeader* thisPHdr = thisTransmit->header;
+            string thisHeaderString;
+            thisPHdr->SerializeToString(&thisHeaderString);
+            _metadataBuf = proto::ProtoHeaderWrap::wrap(thisHeaderString);
+            lock_guard<mutex> streamLock(streamMutex);
+            bool metaSet = _sendChannel->setMetadata(_metadataBuf.data(), _metadataBuf.size());
```

Comment:
I presume my reading of the definition of setMetadata was in the wrong context and this setMetadata is actually calling the SSI version. right?

Reply:
It should be.